File: test/src/go/t05_branch_never.go
package main

import (
	"os"
	"runtime"
	"syscall"
	"time"
	"unsafe"
)

func branchTarget() {
	_, _, _ = syscall.RawSyscall(syscall.SYS_GETTID, 0, 0, 0)
	_, _, _ = syscall.RawSyscall(syscall.SYS_GETUID, 0, 0, 0)
	_, _, _ = syscall.RawSyscall(syscall.SYS_GETGID, 0, 0, 0)
	_, _, _ = syscall.RawSyscall(syscall.SYS_UMASK, 022, 0, 0)
	_, _, _ = syscall.RawSyscall(syscall.SYS_SCHED_YIELD, 0, 0, 0)
}

func main() {
	_ = os.Getpid()
	_ = os.Getppid()

	var u syscall.Utsname
	_ = syscall.Uname(&u)

	time.Sleep(13 * time.Millisecond)
	runtime.Gosched()

	never := false
	if never {
		branchTarget()
	}

	_ = unsafe.Pointer(nil)
}

File: test/src/c/t05_branch_never.c
#include <unistd.h>
#include <sys/syscall.h>
#include <sys/utsname.h>
#include <time.h>
#include <sched.h>

__attribute__((noinline))
static void branch_target(void) {
    (void)syscall(SYS_gettid);
    (void)syscall(SYS_getuid);
    (void)syscall(SYS_getgid);
    (void)syscall(SYS_umask, 022);
    (void)syscall(SYS_sched_yield);
}

int main(void) {
    struct utsname u;
    struct timespec ts;

    volatile int never = 0;

    (void)getpid();
    (void)getppid();
    (void)uname(&u);

    ts.tv_sec = 0;
    ts.tv_nsec = 13 * 1000 * 1000;
    (void)nanosleep(&ts, 0);

    (void)sched_yield();

    if (never) branch_target();
    return 0;
}
